infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/StreamsSnapshotReader.java

```diff
@@ -56,68 +67,103 @@
             //mark the end of the current stream.
             currentMsgTs = snapshotTimestamp;
         }
 
-        ByteBuf buf = Unpooled.buffer();
-        OpaqueEntry.serialize(buf, opaqueEntry);
+        LogReplicationEntry txMsg = new LogReplicationEntry(MessageType.SNAPSHOT_MESSAGE, topologyConfigId, snapshotRequestId, currentMsgTs,
+                preMsgTs, snapshotTimestamp, sequence, opaqueEntry);
 
-        org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntry txMsg = new org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntry
-                (MessageType.SNAPSHOT_MESSAGE, topologyConfigId, snapshotRequestId, currentMsgTs,
-                preMsgTs, snapshotTimestamp, sequence, buf.array());
         preMsgTs = currentMsgTs;
         sequence++;
-        log.debug("Generate TxMsg {}", txMsg.getMetadata());
+
+        log.trace("txMsg {} deepsize sizeInBytes {} entryList.sizeInByres {}  with numEntries {} deepSize sizeInBytes {}",
+                txMsg.getMetadata(), MetricsUtils.sizeOf.deepSizeOf(txMsg), entryList.getSizeInBytes(), entryList.getSmrEntries().size(), MetricsUtils.sizeOf.deepSizeOf(entryList.smrEntries));
+
         return txMsg;
     }
 
     /**
-     * Read numEntries from the current stream.
+     * Read log data from the current stream until the sum of all SMR entries's sizeInBytes reaches the maxDataSizePerMsg.
      * @param stream
-     * @param numEntries
      * @return
      */
-    private List<SMREntry> next(OpaqueStreamIterator stream, int numEntries) {
-        //if it is the end of the stream, set an end of stream mark, the current
-        List<SMREntry> list = new ArrayList<>();
+    private SMREntryList next(OpaqueStreamIterator stream) {
+        List<SMREntry> smrList = new ArrayList<>();
+        int currentMsgSize = 0;
+
         try {
-            while (stream.iterator.hasNext() && list.size() < numEntries) {
-                OpaqueEntry entry = (OpaqueEntry) stream.iterator.next();
-                stream.maxVersion = Math.max(stream.maxVersion, entry.getVersion());
-                list.addAll(entry.getEntries().get(stream.uuid));
+            while (currentMsgSize < maxDataSizePerMsg) {
+                if (lastEntry != null) {
+                    List<SMREntry> smrEntries = lastEntry.getEntries().get(stream.uuid);
+                    if (smrEntries != null) {
+                        int currentEntrySize = ReaderUtility.calculateSize(smrEntries);
+
+                        if (currentEntrySize > MAX_LOG_REPLICATION_DATA_MSG_SIZE_SUPPORTED) {
+                            log.error("The current entry size {} is bigger than the maxDataSizePerMsg {} supported", currentEntrySize, MAX_LOG_REPLICATION_DATA_MSG_SIZE_SUPPORTED);
+                            throw new IllegalSnapshotEntrySizeException(" The snapshot entry is bigger than the system supported");
+                        } else if (currentEntrySize > maxDataSizePerMsg) {
```

Comment:
I'm confused, why do we have this same comparison of a single currentEntrySize twice? shouldn't it only compare with maxDataSizePerMsg, as this is the user configured (which could default to MAX_LOG_REPLICATION_DATA_MSG_SIZE_SUPPORTED)

Reply:
User can setup a different one. Supported is the one we can support. The user setup one is the one the user thinks it gives the best performance. Like 64MB is the max we can support. But in reality maybe 2M gives the best performance. But it depends the real hardware config that the users are aware of or have done experiments with.